README.md

Python package with console commands to list, read, and test NVMe SSD installed in computers running Windows and
Linux OS.

## Features

* Console commands to:
  * List NVMe devices
  * Read NVMe information
  * Check NVMe health
  * Test NVMe features
* python module that can:
  * Read NVMe information
  * Read and check NVMe information at regular interval (e.g. check SMART attributes every second)
  * Check for critical warnings, thermal throttling, excessive wear, etc.
  * Run self-test diagnostic
  * Create simple test and test suites


## Installation

```
pip install nvmetools
```
On Linux OS the nvmecmd utility must be granted access to read NVMe devices with the below commands.  Run
the listnvme console command and it will display the below commands with the actual nvmecmd path.  Run these
commands to grant nvmecmd access to read NVMe devices.
```
    sudo chmod 777 <path to nvmecmd>
    sudo setcap cap_sys_admin,cap_dac_override=ep <path to nvmecmd>
```
<br>

## List NVMe devices

This console command lists the NVMe devices with an unique NVMe number that is required by the other
console commands.
```
listnvme
```

<details>
  <summary>Example console output</summary>

  ```
  Epic! NVMe Tools, version 0.0.7, www.nvmetools.com, Copyright (C) 2022 Joe Jones

  On Window systems the NVMe number is the physical drive number.
  For example, physicaldrive2 would be listed as NVMe 2.

  On Linux systems the NVMe number is the nvme devices number.
  For example, /dev/nvme2 would be listed as NVMe 2.

       LIST OF NVME DRIVES

       NVMe 0 : Sandisk WDC WDS250G2B0C-00PXH0 250GB
       NVMe 1 : Samsung SSD 970 EVO Plus 250GB
  ```
</details>
<br/>

## Read NVMe information

This console command reads the information for an NVMe device.  Provide the NVMe number with the --nvme
parameter to specify the NVMe to read.  For example, the command below reads NVMe 0.
```
readnvme --nvme 0
```
This command displays NVMe information to the console and logs it to a couple of files.  The console output is
logged to readnvme.log and the detailed NVMe information is logged to nvme.info.json.

[Example nvme.info.json](https://github.com/jtjones1001/nvmetools/blob/7c29927faf9bf7dc3a33cfde8fd7c47dd4d78de4/docs/examples/readnvme/nvme.info.json)

By default, only a subset of NVMe parameters are displayed to the console.  All parameters can be displayed
with the -all parameter.  The raw hex data from the commands can be displayed with the --hex parameter.

<details>
  <summary>Example console output</summary>

         ------------------------------------------------------------------------------------------
          NVME DRIVE 0  (/dev/nvme0)
         ------------------------------------------------------------------------------------------
          Vendor                                             Sandisk
          Model Number (MN)                                  WDC WDS250G2B0C-00PXH0
          Serial Number (SN)                                 2035A0805352
          Size                                               250 GB
          Version (VER)                                      1.4.0

          Number of Namespaces (NN)                          1
          Namespace 1 Size                                   250 GB
          Namespace 1 Active LBA Size                        512
          Namespace 1 EUID                                   001b44-8b49bc0ecb
          Namespace 1 NGUID                                  e8238fa6bf530001-001b44-8b49bc0ecb

          Firmware Revision (FR)                             211070WD
          Firmware Slots                                     2
          Firmware Activation Without Reset                  Supported

          Maximum Data Transfer Size (MDTS)                  128
          Enable Host Memory (EHM)                           Enabled
          Host Memory Buffer Size (HSIZE)                    8,192 pages
          Volatile Write Cache (VWC)                         Supported
          Volatile Write Cache Enable (WCE)                  Enabled

          Critical Warnings                                  No
          Media and Data Integrity Errors                    0
          Number Of Failed Self-Tests                        0
          Number of Error Information Log Entries            1

         ----------------------------------------------------------------------
          Temperature       Value          Under Threshold     Over Threshold
         ----------------------------------------------------------------------
          Composite         25 C           -5 C                80 C

         ------------------------------------------------------------------------
          Throttle      Total       TMT1        TMT2        WCTEMP      CCTEMP
         ------------------------------------------------------------------------
          Time (Hrs)    0.850       0.000       0.000       0.014       0.001
          Threshold                 Disabled    Disabled    80 C        85 C
          Count                     0           0           --          --

          Available Spare                                    100 %
          Available Spare Threshold                          10 %
          Controller Busy Time                               15,158 Min
          Data Read                                          339,285.881 GB
          Data Written                                       114,666.719 GB
          Host Read Commands                                 8,937,852,313
          Host Write Commands                                4,997,601,165
          Percentage Used                                    16 %
          Power On Hours                                     1,733
          Power Cycles                                       146
          Unsafe Shutdowns                                   22

         ------------------------------------------------------------------------------------------
          State   NOP    Max         Active      Idle        Entry Latency   Exit Latency
         ------------------------------------------------------------------------------------------
          0              3.5 W       1.8 W       0.63 W
          1              2.4 W       1.6 W       0.63 W
          2              1.9 W       1.5 W       0.63 W
          3       Yes    0.02 W                  0.02 W      3,900 uS        11,000 uS
          4       Yes    0.005 W                 0.005 W     5,000 uS        39,000 uS

          Autonomous Power State Transition                  Supported
          Autonomous Power State Transition Enable (APSTE)   Enabled
          Non-Operational Power State Permissive Mode        Supported
          Non-Operational Power State Permissive Mode Enable (NOPPME) Enabled

          PCI Width                                          x4
          PCI Speed                                          Gen3 8.0GT/s
          PCI Rated Width                                    x4
          PCI Rated Speed                                    Gen3 8.0GT/s

         ------------------------------------------------------------------------------------------
          PCI         Vendor              Vendor ID    Device ID    Location
         ------------------------------------------------------------------------------------------
          Endpoint    Sandisk             0x15B7       0x5009       Bus 1, device 0, function 0
          Root                            0x8086       0xA340       Bus 0, device 27, function 0

</details>

<details>
  <summary>Example console output with --all option</summary>

          128-bit Host Identifier                            Not Supported
          ANA Group Identifier Maximum (ANAGRPMAX)           Not Supported
          ANA Transition Time (ANATT)                        Not Supported
          Abort Command Limit (ACL)                          5
          Admin Vendor Specific command handling             Not Vendor Specific
          Aggregation Threshold (THR)                        1
          Aggregation Time (TIME)                            No Delay
          Arbitration Burst (AB)                             4 (2^4=16)
          Associated Function Type                           PCI
          Asymmetric Namespace Access Change Notices         Not Supported
          Asymmetric Namespace Access Reporting              Not Supported
          Asynchronous Event Request Limit (AERL)            8
          Atomic Write Unit Normal (AWUN)                    1
          Atomic Write Unit Power Fail (AWUPF)               1
          Autonomous Power State Transition                  Supported
          Autonomous Power State Transition Enable (APSTE)   Enabled
          Available Space Below Threshold                    No
          Available Spare                                    100 %
          Available Spare Threshold                          10 %
          Block Erase Sanitize                               Supported
          Command Retry Delay Time 1 (CRDT1)                 0 (0 mS)
          Command Retry Delay Time 2 (CRDT2)                 0 (0 mS)
          Command Retry Delay Time 3 (CRDT3)                 0 (0 mS)
          Commands Supported and Effects Log Page            Supported
          Compare NVM Command                                Supported
          Compare and Write Fused Operation                  Not Supported
          Composite Temperature                              26 C
          Composite Temperature Over Threshold               80 C
          Composite Temperature Under Threshold              -5 C
          Controller Busy Time                               15,158 Min
          Controller ID (CNTLID)                             1
          Controller Type (CNTRLTYPE)                        I/O Controller
          Controller Vendor                                  Sandisk
          Critical Composite Temperature Threshold (CCTEMP)  85 C
          Critical Composite Temperature Time                2 Min
          Critical Warnings                                  No
          Crypto Erase                                       Not Supported
          Crypto Erase Sanitize                              Not Supported
          Current Number Of Errors                           0
          Current Number Of Self-Tests                       20
          Current Power State (PS)                           4
          Current Self-Test Completion                       0
          Current Self-Test Operation                        No Test In Progress
          Data Read                                          339,285.881 GB
          Data Units Read                                    662,667,737
          Data Units Written                                 223,958,435
          Data Written                                       114,666.719 GB
          Dataset Management NVM Command                     Supported
          Deallocated or Unwritten Logical Block Error Enable (DULBE) Disabled
          Device Self-test Command                           Supported
          Directive Send and Directive Receive Commands      Not Supported
          Disable Normal (DN)                                Not Supported
          Doorbell Buffer Config Command                     Not Supported
          EG Available Space Below Threshold                 No
          EG Critical Warnings                               No
          EG Reliability Degraded                            No
          EG in Read Only                                    No
          Enable Host Memory (EHM)                           Enabled
          Endurance Group Event Log Page Change Notices      Not Supported
          Endurance Group Identifier Maximum (ENDGIDMAX)     0
          Endurance Groups                                   Not Supported
          Error Log Page Entries (ELPE)                      256
          Extended Data for Get Log Page                     Supported
          Extended Device Self-test Time (EDSTT)             44 Min
          FRU Globally Unique Identifier (FGUID)             000000-00000000000000000000000000
          Firmware Activation Notices                        Supported
          Firmware Activation Notices Enable                 Enabled
          Firmware Activation Without Reset                  Supported
          Firmware Active Slot                               1
          Firmware Commit and Image Download Commands        Supported
          Firmware Pending Slot                              Not Reported
          Firmware Revision (FR)                             211070WD
          Firmware Slot 1 Read Status                        Read/Write
          Firmware Slot 1 Revision                           211070WD
          Firmware Slot 2 Revision
          Firmware Slots                                     2
          Firmware Update Granularity (FWUG)                 4 KiB
          Format All Namespaces                              Not Supported
          Format NVM Command                                 Supported
          Get LBA Status capability                          Not Supported
          High Priority Weight (HPW)                         1
          Highest Version Detected                           1.4.0
          Host Controlled Thermal Management (HCTMA)         Supported
          Host Memory Buffer Minimum Descriptor Entry Size (HMMINDS) No limitations
          Host Memory Buffer Minimum Size (HMMIN)            823 (3,292 KiB)
          Host Memory Buffer Preferred Size (HMPRE)          51,200 (204,800 KiB)
          Host Memory Buffer Size (HSIZE)                    8,192
          Host Memory Descriptor List Address (HMDLAL)       0x0A028000
          Host Memory Descriptor List Address (HMDLAU)       0x00000001
          Host Memory Descriptor List Entry Count (HMDLEC)   8
          Host Memory Maximum Descriptors Entries (HMMAXD)   8
          Host Read Commands                                 8,937,852,313
          Host Timestamp                                     1,659,225,286,713 mS
          Host Timestamp Decoded                             2022-07-30 16:54:46.713 DST
          Host Write Commands                                4,997,601,165
          IEEE OUI Identifier (IEEE)                         00-1b-44
          Keep Alive Support (KAS)                           Not Supported
          LBA Status Information Notices                     Not Supported
          Low Priority Weight (LPW)                          1
          Maximum Completion Queue Entry Size                4 (2^4=16)
          Maximum Data Transfer Size (MDTS)                  7 (2^7=128)
          Maximum Number Allowed Namespaces (MNAN)           0
          Maximum Outstanding Commands (MAXCMD)              Not Supported
          Maximum Submission Queue Entry Size                6 (2^6=64)
          Maximum Thermal Management Temperature (MXTMT)     85 C
          Maximum Time for Firmware Activation (MTFA)        5,000 mS
          Media and Data Integrity Errors                    0
          Media in Read Only                                 No
          Medium Priority Weight (MPW)                       1
          Minimum Thermal Management Temperature (MNTMT)     0 C
          Model Number (MN)                                  WDC WDS250G2B0C-00PXH0
          NVM Set Identifier Maximum (NSETIDMAX)             0
          NVM Sets                                           Not Supported
          NVM Subsystem Controllers                          Single
          NVM Subsystem NVMe Qualified Name (SUBNQN)         nqn.2018-01.com.wdc:nguid:E8238FA6BF53-0001-001B448B49BC0ECB
          NVM Subsystem PCIe Ports                           Single
          NVME MI Send/Receive Commands                      Not Supported
          Namespace 1 ANA Group Identifier (ANAGRPID)        Not Reported
          Namespace 1 Active LBA Format                      0
          Namespace 1 Atomic Boundary Offset (NABO)          7
          Namespace 1 Atomic Boundary Size Normal (NABSN)    7
          Namespace 1 Atomic Boundary Size Power Fail (NABSPF) 7
          Namespace 1 Atomic Compare & Write Unit (NACWU)    Same as ACWU
          Namespace 1 Atomic Write Unit Normal (NAWUN)       7
          Namespace 1 Atomic Write Unit Power Fail (NAWUPF)  7
          Namespace 1 Atomic Writes                          Supported
          Namespace 1 Capacity (NCAP)                        488,397,168
          Namespace 1 Deallocate Bit in Write Zeros          Supported
          Namespace 1 Deallocate Guard Field                 Not Supported
          Namespace 1 Deallocate Logical Block Value         All 00h
          Namespace 1 Endurance Group Identifier (ENDGID)    Not Supported
          Namespace 1 Exclusive Access All Registrants Reservation Not Supported
          Namespace 1 Exclusive Access Registrants Only Reservation Not Supported
          Namespace 1 Exclusive Access Reservation           Not Supported
          Namespace 1 Extended Data LBA                      Not Supported
          Namespace 1 Format Percent Complete                0
          Namespace 1 Format Progress Indicator              Supported
          Namespace 1 Globally Unique Identifier (NGUID)     e8238fa6bf530001-001b44-8b49bc0ecb
          Namespace 1 IEEE Extended Unique Identifier (EUI64) 001b44-8b49bc0ecb
          Namespace 1 IO Optimize Fields                     Not Supported
          Namespace 1 Ignore Existing Key Specification      1.2.1 or earlier
          Namespace 1 LBA 0 Data Size (LBADS)                9 (2^9=512) *
          Namespace 1 LBA 0 Relative Performance (RP)        Good Performance *
          Namespace 1 LBA 1 Data Size (LBADS)                12 (2^12=4096)
          Namespace 1 LBA 1 Relative Performance (RP)        Better Performance
          Namespace 1 Logical Block Error                    Not Supported
          Namespace 1 Metadata Transfer Buffer               Not Supported
          Namespace 1 Metadata Transfer Extended LBA         Not Supported
          Namespace 1 NGUID/EUID Not Reused                  Not Supported
          Namespace 1 NVM Capacity (NVMCAP)                  250,059,350,016
          Namespace 1 NVM Set Identifier (NVMSETID)          Not Supported
          Namespace 1 Number of LBA Formats (NLBAF)          2
          Namespace 1 Optimal IO Boundary (NOIOB)            Not Reported
          Namespace 1 Optimal Write Size (NOWS)              1
          Namespace 1 Persist Through Power Loss             Not Supported
          Namespace 1 Preferred Deallocate Alignment (NPDA)  1
          Namespace 1 Preferred Deallocate Granularity (NPDG) 1
          Namespace 1 Preferred Write Alignment (NPWA)       1
          Namespace 1 Preferred Write Granularity (NPWG)     1
          Namespace 1 Protection First                       Not Supported
          Namespace 1 Protection Information Enabled         Disabled
          Namespace 1 Protection Information First           Last 8 Bytes
          Namespace 1 Protection Last                        Not Supported
          Namespace 1 Protection Type 1                      Not Supported
          Namespace 1 Protection Type 2                      Not Supported
          Namespace 1 Protection Type 3                      Not Supported
          Namespace 1 Shared                                 Not Supported
          Namespace 1 Size                                   250 GB
          Namespace 1 Size in GiB                            232.9 GiB
          Namespace 1 Size in LBA (NSZE)                     488,397,168
          Namespace 1 Thin Provisioning                      Not Supported
          Namespace 1 Utilization (NUSE)                     488,397,168
          Namespace 1 Write Exclusive All Registrants Reservation Not Supported
          Namespace 1 Write Exclusive Registrants Only Reservation Not Supported
          Namespace 1 Write Exclusive Reservation            Not Supported
          Namespace 1 Write Protected                        No
          Namespace Attribute Notices                        Not Supported
          Namespace Granularity                              Not Supported
          Namespace Management and Attachment Commands       Not Supported
          No-Deallocate Inhibited (NDI)                      Supported
          No-Deallocate Modifies Media After Sanitize (NODMMAS) Media not modified
          Non-Operational Power State Permissive Mode        Supported
          Non-Operational Power State Permissive Mode Enable (NOPPME) Enabled
          Non-zero ANAGRPID                                  Not Supported
          Number Of Failed Self-Tests                        0
          Number of ANA Group Identifiers (NANAGRPID)        Not Supported
          Number of Error Information Log Entries            1
          Number of Namespaces (NN)                          1
          Number of Power States Support (NPSS)              5
          OS Location                                        /dev/nvme0
          One Self-Test                                      Per System
          Overwrite Sanitize                                 Not Supported
          PCI Device ID                                      0x5009
          PCI Location                                       Bus 1, device 0, function 0
          PCI Rated Speed                                    Gen3 8.0GT/s
          PCI Rated Width                                    x4
          PCI Speed                                          Gen3 8.0GT/s
          PCI Subsystem Vendor ID (SSVID)                    0x15B7
          PCI Vendor ID (VID)                                0x15B7
          PCI Width                                          x4
          PCIe Management Endpoint (PCIEME)                  Not Supported
          Percentage Used                                    16 %
          Permanent Write Protect                            Not Supported
          Persistent Event Log                               Supported
          Persistent Event Log Size (PELS)                   64 KiB
          Persistent Memory Unreliable                       No
          Power Cycles                                       146
          Power On Hours                                     1,733
          Power State 0 Active Power (ACTP)                  1.8 Watts
          Power State 0 Active Power Workload (APW)          Workload #2
          Power State 0 Entry Latency (ENLAT)                Not Reported
          Power State 0 Exit Latency (EXLAT)                 Not Reported
          Power State 0 Idle Power (IDLP)                    0.63 Watts
          Power State 0 Idle Time Prior to Transition (ITPT) 100 mS
          Power State 0 Idle Transition Power State (ITPS)   3
          Power State 0 Maximum Power (MP)                   3.5 Watts
          Power State 0 Non-Operational State (NOPS)         False
          Power State 0 Relative Read Latency (RRL)          0
          Power State 0 Relative Read Throughput (RRT)       0
          Power State 0 Relative Write Latency (RWL)         0
          Power State 0 Relative Write Throughput (RWT)      0

CHANGELOG.md
# Changelog

## v0.8.0

### Improvements
- Overhauled the test framework
- Sped up the creation of PDF report

## v0.7.0

### Improvements
- Overhauled the html report from viewnvme/testnvme/checknvme
- Updated parameters relevant to NVMe health

### For developers
- Updated nvmecmd version with parameter changes for health and bug fix for feature reads

## v0.6.0

### Improvements
- Removed pycache and hidden OS files from the build
- Updated nvmecmd to version that supports sanitize
- Changed testnvme so user can easily add their own test suites
- Updated big file write test case

### For developers
- Split out test suites into separate files

## v0.5.0

 Added viewnvme command.  Several bug fixes and improvements in the test cases and reporting.

### Improvements
- Added viewnvme command
- Improvements and bug fixes in the html resources
- Bug fixes in the test cases, reporting, and framework


## v0.4.3

 Updated to support PCIe gen4/5 devices, several bug fixes in framework to stop on fail

### Improvements
- Updated nvmecmd to version that supports PCIe gen4/5 speed
- Set read default to exclude log pages 7/8 because they can be too big
- Bug fixes in framework for stop and stop on fail options

## v0.4.2

Added fixes for bugs found through initial testing and updated test cases to be more consistent and
simple

### Improvements
- Replaced aspm_latency and nonop_latency test cases with idle_latency
- Test Case improvements to make them more consistent and simpler
- Bug fix for issue where compare fails when new parameters added during a test suite
- Added check for free disk space in test suites that require full drive testing

## v0.4.1

Initial public release of new format.  This is a beta release with limited testing and is posted
primarily to gather feedback.

### Improvements
- Added test framework for testing NVMe devices
- Added html dashboard of test results

### For developers
- The prior release 0.2.1 was renamed to nvmetools-original-version and is available on github



src/nvmetools/apps/nvmecmd.py
# --------------------------------------------------------------------------------------
# Copyright(c) 2023 Joseph Jones,  MIT License @  https://opensource.org/licenses/MIT
# --------------------------------------------------------------------------------------
r"""Define functions for using nvmecmd utility.

nvmecmd uses admin commands to communicate with nvme device
some admin commands, like self-test, require admin rights
nvmecmd is included with the package in <package>\resources\nvmecmd
nvmecmd requires output directory to be there before it runs
nvmecmd -v input enables verbose logging

//------------------------------------------------------------------------------
// Exit codes for nvmecmd from C++ project
//------------------------------------------------------------------------------
inline const int NVMECMD_USAGE_ERROR_CODE = 16;
inline const int NVMECMD_EXCEPTION_CODE   = 17;
inline const int NVMECMD_NO_NVME_DRIVES = 18;
inline const int NVMECMD_NVME_DRIVE_NOT_FOUND = 19;

inline const int READINFO_FIRST_SAMPLE_READ_ERROR = 20;
inline const int READINFO_FAIL_SAMPLES = 21;
inline const int READINFO_FAIL_LIMIT = 22;

inline const int SELFTEST_NOT_SUPPORTED_ERROR = 30;
inline const int SELFTEST_NOT_STARTED_ERROR = 31;
inline const int SELFTEST_NO_EDSTT_ERROR = 32;
inline const int SELFTEST_FAILED = 33;
inline const int SELFTEST_ABORT_FAILED_ERROR = 34;
inline const int SELFTEST_WRONG_STATUS_ERROR = 35;
inline const int SELFTEST_ABORT_READ_ERROR = 36;
"""
import json
import os
import platform

from nvmetools import SRC_DIRECTORY
from nvmetools.support.conversions import (
    MS_IN_MIN,
    MS_IN_SEC,
    SEC_IN_MIN,
    as_datetime,
    as_int,
    as_linear,
    as_monotonic,
    is_debug,
)
from nvmetools.support.process import RunProcess

LINEAR_LIMIT = 0.9

NVMECMD_DIR = os.path.join(SRC_DIRECTORY, "nvmetools", "resources", "nvmecmd")

if "Windows" == platform.system():
    NVMECMD_EXEC = os.path.join(NVMECMD_DIR, "nvmecmd.exe")
else:
    NVMECMD_EXEC = os.path.join(NVMECMD_DIR, "nvmecmd")

READ_INFO_FILE = "nvme.info.json"
FIRST_SAMPLE_READ_FILE = "nvme.info.sample-1.json"
SELFTEST_FILE = "selftest.summary.json"
READ_SUM_FILE = "read.summary.json"
TRACE_LOG = "nvmecmd.trace.log"
NVMECMD_READ_CMD = os.path.join(NVMECMD_DIR, "read.cmd.json")
NVMECMD_SELFTEST_CMD = os.path.join(NVMECMD_DIR, "self-test.cmd.json")
NVMECMD_EXT_SELFTEST_CMD = os.path.join(NVMECMD_DIR, "extended-self-test.cmd.json")

# Use exception code 51-55 for this module


class _NoNvme(Exception):
    def __init__(self, nvme):
        self.code = 51
        self.nvmetools = True
        super().__init__(f" NVME device {nvme} was not found.")


class _NvmecmdBadJson(Exception):
    def __init__(self, error, file):
        self.code = 52
        self.nvmetools = True
        error_msg = f"{file} @line {error.lineno} @col {error.colno}"
        super().__init__(f" failed parsing nvmecmd JSON file {error_msg}.")


class _NvmecmdException(Exception):
    def __init__(self, directory):
        self.code = 53
        self.nvmetools = True
        tracelog = os.path.join(directory, TRACE_LOG)
        super().__init__(f" nvmecmd had exception, see {tracelog}.")


class NvmecmdPermissionError(Exception):
    """Custom exception to indicate nvmecmd missing permission."""

    def __init__(self):
        """Initialize exception."""
        self.code = 54
        self.nvmetools = True
        error_msg = " nvmecmd utility does not have permission to access NVMe.\n\n"
        error_msg += " To give permission run these commands:\n\n"
        error_msg += f"  sudo chmod 777 {NVMECMD_EXEC} \n"
        error_msg += f"  sudo setcap cap_sys_admin,cap_dac_override=ep {NVMECMD_EXEC} \n"
        super().__init__(error_msg)


def check_nvmecmd_permissions():
    """Check nvmecmd permission to read NVMe device."""
    if "Windows" != platform.system():
        attribute = "security.capability"
        expected_value = "0100000202002000000000000000000000000000"
        try:
            current_value = os.getxattr(NVMECMD_EXEC, attribute).hex()
            if current_value != expected_value:
                raise NvmecmdPermissionError()
        except OSError:
            raise NvmecmdPermissionError()

        if int(oct(os.stat(NVMECMD_EXEC).st_mode)[-3:]) != 777:
            raise NvmecmdPermissionError()

    # TODO: Check check_nvmecmd_permissions on other versions of Linux, only tested on Fedora


class Read:
    """Read NVMe information using nvmecmd utility.

    Attributes:
       info (json): Information read from nvme.info.json.
       return_code (int): return code.
       run_time (float): Time process ran in seconds.
       summary (json): Summary information from read.summary.json.
    """

    def __init__(
        self,
        nvme=0,
        directory=".",
        samples=1,
        interval=0,
        cmd_file="read",
        wait=True,
    ):
        """Start reading information.

        Args:
           nvme: NVMe device number.
           directory: Directory to log results.
           samples: Number of samples to read.
           intervalerval between samples in mS.
           cmd_file: Name of cmd file to use for the read.
           wait:  Waits for read to complete if True.
        """
        cmd_file_path = os.path.join(NVMECMD_DIR, f"{cmd_file}.cmd.json")
        self._samples = samples
        self._directory = os.path.abspath(directory)
        self._nvme = nvme

        check_nvmecmd_permissions()

        args = [
            NVMECMD_EXEC,
            cmd_file_path,
            "--dir",
            f"{self._directory}",
            "--nvme",
            f"{self._nvme}",
            "--samples",
            f"{self._samples}",
            "--interval",
            f"{interval}",
        ]
        if is_debug():
            args.append("-v")

        if interval == 0:
            self._timeout = 60 + samples
        else:
            self._timeout = 60 + samples * interval / MS_IN_SEC

        self._process = RunProcess(args, directory, wait=False)
        if wait:
            self.wait()

    def wait(self):
        """Wait for information to be ready.

        Returns:
           Return code from nvmecmd process.

        Raises:
           _NvmecmdBadJson if the info or summary file not correctly formatted.
        """
        self._process.wait(self._timeout)
        self.return_code = self._process.return_code
        self.run_time = self._process.run_time

        if self.return_code in [16, 17]:
            raise _NvmecmdException(self._directory)
        if self.return_code in [18, 19]:
            raise _NoNvme(self._nvme)

        info_file = FIRST_SAMPLE_READ_FILE if self._samples > 1 else READ_INFO_FILE

        try:
            info_file = os.path.join(self._directory, info_file)
            with open(info_file, "r") as file_object:
                self.info = json.load(file_object)

            summary_file = os.path.join(self._directory, READ_SUM_FILE)
            with open(summary_file, "r") as file_object:
                self.summary = json.load(file_object)

        except json.JSONDecodeError as error:
            raise _NvmecmdBadJson(error, summary_file)

        if not is_debug() and self.return_code == 0:
            tracelog = os.path.join(self._directory, TRACE_LOG)
            os.remove(tracelog)

        return self.return_code

    def stop(self):
        """Stop reading information gracefully.

        Stops the read when doing multiple samples by stopping the nvmecmd process.
        """
        self._process.stop()


class Selftest:
    """Run drive self-test diagnostic.

    Runs the short or extended self-test diagnostic.

    Attributes:
       data (dictionary): Dictionary of parameters
    """

    def __init__(self, nvme, directory, extended=False, limit_min=2):
        """Run self-test diagnostic.

        Args:
            nvme:  nvme device number.
            directory: Directory where results are logged.
            extended: Runs extended self-test if True, else the short self-test.
            limit_min: Time limit for self-test in minutes.
        """
        self._directory = directory
        self._nvme = nvme
        self._extended = extended
        self.data = {}
        self.data["runtime limit"] = limit_min

        if extended:
            cmd_file = NVMECMD_EXT_SELFTEST_CMD
        else:
            cmd_file = NVMECMD_SELFTEST_CMD

        check_nvmecmd_permissions()

        args = [
            NVMECMD_EXEC,
            cmd_file,
            "--dir",
            f"{directory}",
            "--nvme",
            f"{nvme}",
        ]
        if is_debug():
            args.append("-v")

        self.data["return code"] = RunProcess(args, directory).return_code

        if self.data["return code"] in [16, 17]:
            raise _NvmecmdException(directory)
        if self.data["return code"] in [18, 19]:
            raise _NoNvme(nvme)
        if self.data["return code"] not in [30, 31, 32]:
            try:
                info_file = os.path.join(directory, SELFTEST_FILE)
                with open(info_file, "r") as file_object:
                    self.data["logfile"] = json.load(file_object)
            except json.JSONDecodeError as error:
                raise _NvmecmdBadJson(error, info_file)

            elapsed_time = []
            elapsed_progress_time = [0]  # first sample has status 0 but want to start at 0,0
            elapsed_progress = [0]

            details = self.data["logfile"]["selftest details"]
            self.data["runtime"] = details["run time in ms"] / MS_IN_MIN
            start = as_datetime(details["status"][0]["timestamp"])

            for sample in details["status"]:
                delta = as_datetime(sample["timestamp"]) - start
                elapsed_time.append(delta.total_seconds() / SEC_IN_MIN)
                if sample["status"] != 0:
                    elapsed_progress.append(int(sample["percent complete"]))
                    elapsed_progress_time.append(delta.total_seconds() / SEC_IN_MIN)

            self.data["monotonic"] = as_monotonic(elapsed_progress)
            self.data["linear"] = as_linear(elapsed_progress_time, elapsed_progress)
            self.data["result_poh"] = as_int(details["log page 6 power on hours"])
            self.data["last_poh"] = as_int(details["status"][-1]["Power On Hours"])
            self.data["second_last_poh"] = as_int(details["status"][-2]["Power On Hours"])

        if not is_debug() and self.data["return code"] == 0:
            tracelog = os.path.join(directory, TRACE_LOG)
            os.remove(tracelog)
